holoviews/plotting/mpl/element.py

```diff
@@ -637,8 +640,13 @@ class OverlayPlot(LegendPlot, GenericOverlayPlot):
     _passed_handles = ['fig', 'axis']
 
     def __init__(self, overlay, ranges=None, **params):
-        if overlay.traverse(lambda x: x, (Element3D,)):
-            params['projection'] = '3d'
+        if 'projection' not in params:
+            projs = self._deep_options(overlay, 'plot', ['projection'],
+                                       [Element])['projection']
+            if len(set(projs)) > 1:
+                raise Exception("A single axis may only be assigned one projection type")
```

Comment:
No you had a point, it was also calling traverse twice which is unnecessary so I've made ``traverse_options`` a bit more powerful and factored out a ``_get_projection`` method.